canonical-web-and-design/maas-ui
# fix(ui): fix KVM details resource usage calculations

## Done

- Fixed resource calculations in KVM details pages, taking over-commit ratio into account

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- This is probably easiest to QA on bolla. Make sure the `bolla` and `other` KVMs have over-commit ratios for both CPU and Memory set to 1.
- Go to the Resources tab for each LXD VM host and check that the "Overall" card makes sense, i.e. the "Other" value for one is the "Project" value for the other (since there are only two projects in that LXD host)
- Check that the Project Resources card values make sense, i.e. the values for "Allocated" and "Free" correspond to the "Project" and "Free" values in the "Overall" card
- Check the Project Summary card values make sense (they should be identical to the Project Resources card)
- Compose a VM in `bolla` and make sure that it just exceeds the free amount for both memory and cores.
- Check that all the RAM and Cores charts show up as yellow because the resources are over-committed.
- Change `bolla`'s over-commit ratio for both memory and CPU to 2.
- Check that the RAM and core charts are no longer yellow.
- Go to the KVM details page for `other`. It should be the same as before because it doesn't know about the newly composed VM on `bolla` yet.
- Refresh the KVM. This is an expensive operation and only happens on a cron normally, but we want the latest data NOW.
- Check that the charts are all yellow.
- Change `other`'s over-commit ratio for both memory and CPU to 2.
- Check that the RAM and core charts are no longer yellow.

## Fixes

Fixes #2560
